pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -0,0 +1,981 @@
+        -------
+        Timestamp with fields replaced
+        """
+
+        cdef:
+            pandas_datetimestruct dts
+            int64_t value, value_tz, offset
+            object _tzinfo, result, k, v
+            datetime ts_input
+
+        # set to naive if needed
+        _tzinfo = self.tzinfo
+        value = self.value
+        if _tzinfo is not None:
+            value_tz = tz_convert_single(value, _tzinfo, 'UTC')
+            value += value - value_tz
+
+        # setup components
+        dt64_to_dtstruct(value, &dts)
+        dts.ps = self.nanosecond * 1000
+
+        # replace
+        def validate(k, v):
+            """ validate integers """
+            if not is_integer_object(v):
+                raise ValueError("value must be an integer, received "
+                                 "{v} for {k}".format(v=type(v), k=k))
+            return v
+
+        if year is not None:
+            dts.year = validate('year', year)
+        if month is not None:
+            dts.month = validate('month', month)
+        if day is not None:
+            dts.day = validate('day', day)
+        if hour is not None:
+            dts.hour = validate('hour', hour)
+        if minute is not None:
+            dts.min = validate('minute', minute)
+        if second is not None:
+            dts.sec = validate('second', second)
+        if microsecond is not None:
+            dts.us = validate('microsecond', microsecond)
+        if nanosecond is not None:
+            dts.ps = validate('nanosecond', nanosecond) * 1000
+        if tzinfo is not object:
+            _tzinfo = tzinfo
+
+        # reconstruct & check bounds
+        ts_input = datetime(dts.year, dts.month, dts.day, dts.hour, dts.min,
+                            dts.sec, dts.us, tzinfo=_tzinfo)
+        ts = convert_datetime_to_tsobject(ts_input, _tzinfo)
+        value = ts.value + (dts.ps // 1000)
+        if value != NPY_NAT:
+            check_dts_bounds(&dts)
+
+        return create_timestamp_from_ts(value, dts, _tzinfo, self.freq)
+
+    def isoformat(self, sep='T'):
+        base = super(_Timestamp, self).isoformat(sep=sep)
+        if self.nanosecond == 0:
+            return base
+
+        if self.tzinfo is not None:
+            base1, base2 = base[:-6], base[-6:]
+        else:
+            base1, base2 = base, ""
+
+        if self.microsecond != 0:
+            base1 += "%.3d" % self.nanosecond
+        else:
+            base1 += ".%.9d" % self.nanosecond
+
+        return base1 + base2
+
+    def _has_time_component(self):
+        """
+        Returns if the Timestamp has a time component
+        in addition to the date part
+        """
+        return (self.time() != _zero_time
+                or self.tzinfo is not None
+                or self.nanosecond != 0)
+
+    def to_julian_date(self):
+        """
+        Convert TimeStamp to a Julian Date.
+        0 Julian date is noon January 1, 4713 BC.
+        """
+        year = self.year
```

Comment:
can cdef these (TODO)

Reply:
Yah, in fact a bunch of stuff can be moved up to `_Timestamp`.